src/lib/prepareMetadata.ts

```diff
@@ -0,0 +1,292 @@
+    IEA?: string;
+  };
+};
+
+type FunctionalGroup = {
+  release: string;
+  controlNumber: number;
+  applicationSenderCode: string;
+  applicationReceiverCode: string;
+  transactionSets: TransactionSet[];
+  segments: {
+    GS: string;
+    GE?: string;
+  };
+};
+
+type TransactionSet = {
+  controlNumber: number;
+  transactionSetType: string;
+  segments: string[];
+};
+
+// split input into multiple EDI documents, along with metadata:
+//
+export const prepareMetadata = (ediDocument: string): EDIMetadata[] => {
+  const { segmentDelimiter, elementDelimiter } = extractDelimiters(ediDocument);
+
+  // If segmentDelimiter is not a newline, remove trailing newlines
+  if (!segmentDelimiter.match(/(\r?\n|\r)$/g)) {
+    ediDocument = ediDocument.replace(/\r?\n|\r/g, "");
+  }
+
+  // split input on segment delimiter (and filter out any empty segments, such as after the final segment terminator)
+  const segments = ediDocument
+    .split(segmentDelimiter)
+    .filter((segment) => segment);
+
+  let currentInterchange: Interchange | undefined;
+  let currentFunctionalGroup: FunctionalGroup | undefined;
+  let currentTransactionSet: TransactionSet | undefined;
+
+  return segments.reduce((results: EDIMetadata[], currentSegment) => {
+    const elements = currentSegment.trim().split(elementDelimiter);
+    switch (elements[0]) {
+      case "ISA":
+        // this use case could be accommodated in the future by auto-generating an IEA
+        if (currentFunctionalGroup)
+          throw new Error(
+            "interchange start encountered without previous interchange termination"
+          );
+
+        currentInterchange = {
+          ...extractIsaIds(elements),
+          functionalGroups: [],
+          controlNumber: parseInt(elements[12]),
+          delimiters: {
+            element: elementDelimiter,
+            segment: segmentDelimiter,
+          },
+          segments: { ISA: currentSegment },
+        };
+        break;
+      case "GS":
+        if (currentFunctionalGroup) {
+          throw new Error(
+            "new functional group encountered without previous functional group termination"
+          );
+        }
+
+        const release = extractRelease(elements);
+        currentFunctionalGroup = {
+          release,
+          controlNumber: parseInt(elements[5]),
+          applicationSenderCode: elements[2],
+          applicationReceiverCode: elements[3],
+          transactionSets: [],
+          segments: { GS: currentSegment },
+        };
+        break;
+      case "ST":
+        if (currentTransactionSet !== undefined)
+          throw new Error("incomplete transaction set found");
+
+        const transactionSetType = extractTransactionSetType(elements);
+
+        currentTransactionSet = {
+          transactionSetType,
+          controlNumber: parseInt(elements[2]),
+          segments: [currentSegment],
+        };
+
+        break;
+      case "SE":
+        if (!currentFunctionalGroup)
```

Comment:
oh, related to the comment above, looks like maybe this just needs to be moved up